File: qt/gen_qdial.go
package qt

import (
	"unsafe"
)

type QDial struct {
	h          uintptr
	isSubclass bool
}

// NewQDial constructs a new QDial object.
func NewQDial(parent *QWidget) *QDial {
	g := newQDial(QDial_new(parent.cPointer()))
	g.isSubclass = true
	return g
}

// NewQDial2 constructs a new QDial object.
func NewQDial2() *QDial {
	g := newQDial(QDial_new2())
	g.isSubclass = true
	return g
}

func (this *QDial) MetaObject() *QMetaObject {
	return newQMetaObject(QDial_MetaObject(this.h))
}

func (this *QDial) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QDial_Metacast(this.h, param1_Cstring))
}

func QDial_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QDial_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDial) Wrapping() bool {
	return (bool)(QDial_Wrapping(this.h))
}

func (this *QDial) NotchSize() int {
	return (int)(QDial_NotchSize(this.h))
}

func (this *QDial) SetNotchTarget(target float64) {
	QDial_SetNotchTarget(this.h, (double)(target))
}

func (this *QDial) NotchTarget() float64 {
	return (float64)(QDial_NotchTarget(this.h))
}

func (this *QDial) NotchesVisible() bool {
	return (bool)(QDial_NotchesVisible(this.h))
}

func (this *QDial) SizeHint() *QSize {
	_goptr := newQSize(QDial_SizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDial) MinimumSizeHint() *QSize {
	_goptr := newQSize(QDial_MinimumSizeHint(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QDial) SetNotchesVisible(visible bool) {
	QDial_SetNotchesVisible(this.h, (bool)(visible))
}

func (this *QDial) SetWrapping(on bool) {
	QDial_SetWrapping(this.h, (bool)(on))
}

func QDial_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QDial_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QDial_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QDial_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QDial) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QDial_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QDial) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QDial_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDial_MetaObject
func miqt_exec_callback_QDial_MetaObject(self QDial, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QDial{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QDial) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QDial_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QDial) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QDial_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QDial_Metacast
func miqt_exec_callback_QDial_Metacast(self QDial, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QDial{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qdial.cpp
// +build ignore

#include <QAbstractSlider>
#include <QDial>
#include <QEvent>
#include <QMetaObject>
#include <QMouseEvent>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QResizeEvent>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyleOptionSlider>
#include <QWidget>
#include <qdial.h>
#include "gen_qdial.h"

class MiqtVirtualQDial : public virtual QDial {
public:

	MiqtVirtualQDial(QWidget* parent): QDial(parent) {};
	MiqtVirtualQDial(): QDial() {};

	virtual ~MiqtVirtualQDial() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QDial::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QDial_MetaObject(const_cast<MiqtVirtualQDial*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QDial::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QDial::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QDial_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QDial::qt_metacast(param1);

	}

};

QDial* QDial_new(QWidget* parent) {
	return new MiqtVirtualQDial(parent);
}

QDial* QDial_new2() {
	return new MiqtVirtualQDial();
}

void QDial_virtbase(QDial* src, QAbstractSlider** outptr_QAbstractSlider) {
	*outptr_QAbstractSlider = static_cast<QAbstractSlider*>(src);
}

QMetaObject* QDial_MetaObject(const QDial* self) {
	return (QMetaObject*) self->metaObject();
}

void* QDial_Metacast(QDial* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QDial_Tr(const char* s) {
	QString _ret = QDial::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QDial_Wrapping(const QDial* self) {
	return self->wrapping();
}

int QDial_NotchSize(const QDial* self) {
	return self->notchSize();
}

void QDial_SetNotchTarget(QDial* self, double target) {
	self->setNotchTarget(static_cast<double>(target));
}

double QDial_NotchTarget(const QDial* self) {
	qreal _ret = self->notchTarget();
	return static_cast<double>(_ret);
}

bool QDial_NotchesVisible(const QDial* self) {
	return self->notchesVisible();
}

QSize* QDial_SizeHint(const QDial* self) {
	return new QSize(self->sizeHint());
}

QSize* QDial_MinimumSizeHint(const QDial* self) {
	return new QSize(self->minimumSizeHint());
}

void QDial_SetNotchesVisible(QDial* self, bool visible) {
	self->setNotchesVisible(visible);
}

void QDial_SetWrapping(QDial* self, bool on) {
	self->setWrapping(on);
}

struct miqt_string QDial_Tr2(const char* s, const char* c) {
	QString _ret = QDial::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QDial_Tr3(const char* s, const char* c, int n) {
	QString _ret = QDial::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QDial_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQDial*>( (QDial*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QDial_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQDial*)(self) )->virtualbase_MetaObject();
}

void QDial_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQDial*>( (QDial*)(self) )->handle__Metacast = slot;
}

void* QDial_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQDial*)(self) )->virtualbase_Metacast(param1);
}

void QDial_Delete(QDial* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQDial*>( self );
	} else {
		delete self;
	}
}

